src/parquet/encodings/plain-encoding-test.cc

```diff
@@ -63,10 +63,281 @@ TEST(BooleanTest, TestEncodeDecode) {
   ASSERT_EQ(nvalues, values_decoded);
 
   for (size_t i = 0; i < nvalues; ++i) {
-    ASSERT_EQ(BitUtil::GetArrayBit(decode_data, i), draws[i]) << i;
+    ASSERT_EQ(draws[i], BitUtil::GetArrayBit(decode_data, i)) << i;
   }
 }
 
-} // namespace test
+TEST(BooleanTest2, TestEncodeDecode) {
+  size_t nvalues = 10000;
+  size_t nbytes = BitUtil::Ceil(nvalues, 8);
+
+  vector<uint8_t> draws;
+  draws.resize(nvalues);
+  bool* bool_data = reinterpret_cast<bool*>(draws.data());
+  // seed the prng so failure is deterministic
+  random_bools(nvalues, 0.5, bool_data);
 
+  PlainEncoder<Type::BOOLEAN> encoder(nullptr);
+  PlainDecoder<Type::BOOLEAN> decoder(nullptr);
+
+  InMemoryOutputStream dst;
+  encoder.Encode(bool_data, nvalues, &dst);
+
+  std::vector<uint8_t> encode_buffer;
+  dst.Transfer(&encode_buffer);
+
+  ASSERT_EQ(nbytes, encode_buffer.size());
+
+  std::vector<uint8_t> decode_buffer(nvalues);
+  const uint8_t* decode_data = &decode_buffer[0];
+
+  decoder.SetData(nvalues, &encode_buffer[0], encode_buffer.size());
+  size_t values_decoded = decoder.Decode(
+      reinterpret_cast<bool*>(decode_buffer.data()), nvalues);
+  ASSERT_EQ(nvalues, values_decoded);
+
+  for (size_t i = 0; i < nvalues; ++i) {
+    ASSERT_EQ(draws[i], decode_data[i]) << i;
+  }
+}
+
+TEST(Int32Test, TestEncodeDecode) {
+  size_t nvalues = 10000;
+  size_t nbytes = nvalues * type_traits<Type::INT32>::value_byte_size;
+
+  vector<int32_t> draws;
+  // seed the prng so failure is deterministic
+  random_int_numbers(nvalues, 0.5, &draws);
+
+  PlainEncoder<Type::INT32> encoder(nullptr);
+  PlainDecoder<Type::INT32> decoder(nullptr);
+
+  InMemoryOutputStream dst;
+  encoder.Encode(draws.data(), nvalues, &dst);
+
+  std::vector<uint8_t> encode_buffer;
+  dst.Transfer(&encode_buffer);
+
+  ASSERT_EQ(nbytes, encode_buffer.size());
+
+  std::vector<uint8_t> decode_buffer(nbytes);
+  const int32_t* decode_data = reinterpret_cast<int32_t*>(decode_buffer.data());
+
+  decoder.SetData(nvalues, &encode_buffer[0], encode_buffer.size());
+  size_t values_decoded = decoder.Decode(
+      reinterpret_cast<int32_t*>(decode_buffer.data()), nvalues);
+  ASSERT_EQ(nvalues, values_decoded);
+
+  for (size_t i = 0; i < nvalues; ++i) {
+    ASSERT_EQ(draws[i], decode_data[i]) << i;
+  }
+}
+
+TEST(Int64Test, TestEncodeDecode) {
```

Comment:
Any possibility of using templates to reduce code duplication between this and the other primitive type functions?